### PR TITLE
Adding `lovely-tensors` and `lovely-jax` from PyPI

### PR DESCRIPTION
**Pull request by sugatoray**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by sugatoray** [comment]
## Reason for Build Failure: `1/N` | `lovely-jax`

`Observation`:

> :point_right: *`lovely-jax` is missing dependency `matplotlib`*

- `line number: 3177`: [(error)](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=615903&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=3177)
- `time elapsed`: `4m:35s`

```sh
ModuleNotFoundError: No module named 'matplotlib'
```


<details>
<summary><b>Click to expand</b>: Detailed Error Trace</summary>
<p>

```sh
import: 'lovely_jax'
Traceback (most recent call last):
  File "/home/conda/staged-recipes/build_artifacts/lovely-jax_1668990078694/test_tmp/run_test.py", line 2, in <module>
    import lovely_jax
  File "/home/conda/staged-recipes/build_artifacts/lovely-jax_1668990078694/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.11/site-packages/lovely_jax/__init__.py", line 4, in <module>
    from lovely_jax.patch import *
  File "/home/conda/staged-recipes/build_artifacts/lovely-jax_1668990078694/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.11/site-packages/lovely_jax/patch.py", line 9, in <module>
    import matplotlib.pyplot as plt
ModuleNotFoundError: No module named 'matplotlib'
Tests failed for lovely-jax-0.0.1-pyhd8ed1ab_0.conda - moving package to /home/conda/staged-recipes/build_artifacts/broken
WARNING:conda_build.build:Tests failed for lovely-jax-0.0.1-pyhd8ed1ab_0.conda - moving package to /home/conda/staged-recipes/build_artifacts/broken
TESTS FAILED: lovely-jax-0.0.1-pyhd8ed1ab_0.conda
##[error]Bash exited with code '1'.
```

</p>
</details>

**Comment by sugatoray** [comment]
## Reason for Build Failure: `2/N` | `lovely-jax` | `windows`

`Observation`:

> :point_right: *`jaxlib` is not available on conda-forge for windows -- necessary for `lovely-jax`*

- `line number: 2397`: [(error)](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=615918&view=logs&j=240f1fee-52bc-5498-a14a-8361bde76ba0&t=2fb5b0a7-737c-5f5c-8f3f-f6db6174bacf&l=2397)
- `time elapsed`: `6m:47s`

```sh
The reported errors are:
- Encountered problems while solving:
-   - nothing provides jaxlib needed by jax-0.1.39-pyh5ca1d4c_0
```


<details>
<summary><b>Click to expand</b>: Detailed Error Trace</summary>
<p>

```sh
The reported errors are:
- Encountered problems while solving:
-   - nothing provides jaxlib needed by jax-0.1.39-pyh5ca1d4c_0
- 

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "D:\a\1\s\.ci_support\build_all.py", line 220, in <module>
    build_all(os.path.join(root_dir, "recipes"), args.arch)
  File "D:\a\1\s\.ci_support\build_all.py", line 118, in build_all
    build_folders(recipes_dir, folders, arch, channel_urls)
  File "D:\a\1\s\.ci_support\build_all.py", line 176, in build_folders
    conda_build.api.build([recipe], config=get_config(arch, channel_urls))
  File "C:\Miniforge\lib\site-packages\conda_build\api.py", line 186, in build
    return build_tree(
  File "C:\Miniforge\lib\site-packages\conda_build\build.py", line 3091, in build_tree
    packages_from_this = build(metadata, stats,
  File "C:\Miniforge\lib\site-packages\conda_build\build.py", line 2131, in build
    create_build_envs(top_level_pkg, notest)
  File "C:\Miniforge\lib\site-packages\conda_build\build.py", line 2009, in create_build_envs
    raise e
  File "C:\Miniforge\lib\site-packages\conda_build\build.py", line 1988, in create_build_envs
    environ.get_install_actions(m.config.test_prefix,
  File "C:\Miniforge\lib\site-packages\boa\cli\mambabuild.py", line 150, in mamba_get_install_actions
    raise err
conda_build.exceptions.DependencyNeedsBuildingError: Unsatisfiable dependencies for platform win-64: {MatchSpec("jax==0.1.39=pyh5ca1d4c_0"), MatchSpec("jaxlib")}
##[error]Cmd.exe exited with code '1'.
```

</p>
</details>

image

**Comment by sugatoray** [comment]
PR is ready. Request for review/merging.

cc: @conda-forge/help-python 

### Note:

- [x] `lovely-tensors`: passes `linux`, `osx` and `windows` builds.
- [x] `lovely-numpy`: passes `linux`, `osx` and `windows` builds.
- [x] `lovely-jax`: passes `linux`, `osx` builds; but FAILS `windows` build (as `jaxlib` dependency is not available on conda-forge for windows).

**Comment by xl0** [comment]
Heya. Author here. Please wait a bit with -numpy and especially -jax. Both are very much WIP. -tensors should be good to go.

**Comment by sugatoray** [comment]
@xl0 Thank you, for the update. Even if -numpy and -jax are WIP, we can put them on conda-forge and keep updating later on. 

If it’s okay with you, I can add you as a co-maintainer of the conda-forge recipes of the libraries. 

This will help ensure that the libraries are kept up to date on both PyPI and conda-forge, after each release.

**Comment by xl0** [comment]
To be completely honest, I would prefer to wait it out until I have time to put it on conda-forge and maintain it myself.
I'm using nbdev for development, which has support for releasing on conda; I need to find time to set it up.

This should not take very long. For now, let people use pip.

The progress on Lovely Tensors is moving very quickly, I would not want to have conda-forge packages that are outdated, as it happens occasionally. The project is not even 2 months old; let it mature a bit, I'll find time to put it on forge within the next month.

File: recipes/lovely-jax/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+requirements:
+  host:
+    - python >=3.7
+    - pip
+  run:
+    - python >=3.7
+    - fastcore
+    - jax
+    # missing dependency
+    - matplotlib
```
Comment (xhochy) [inline recipes/lovely-jax/meta.yaml:26]: The base variant (no Qt) is typically enough.
```suggestion
    - matplotlib-base
```

**Comment by xhochy** [comment]
These packages are on PyPI, so I don't see an issue with them being on conda-forge. If you don't want them used, why are they then on PyPI at all?

Given that they are pure Python packages, we should be able to auto-update them with grayskull and bot-automerge, i.e. no need for human involvement at all in the maintenance.

**Comment by sugatoray** [comment]
> To be completely honest, I would prefer to wait it out until I have time to put it on conda-forge and maintain it myself.
> 
> I'm using nbdev for development, which has support for releasing on conda; I need to find time to set it up.
> 
> 
> 
> This should not take very long. For now, let people use pip.
> 
> 
> 
> The progress on Lovely Tensors is moving very quickly, I would not want to have conda-forge packages that are outdated, as it happens occasionally. The project is not even 2 months old; let it mature a bit, I'll find time to put it on forge within the next month.

@xl0 

Sounds good. I will remove the other two then. And if it's okay with you, I would like to add you as a co-maintainer. Please let me know.

**Comment by xl0** [comment]
Hii. Sorry for the delay, was busy. I'm looking into this now.

`lovely-tensors` now depends on `lovely-numpy`. Does it mean we should add -numpy first, then -tensors?

For the workflow, as I understand it, I'll have to run `nbdev_conda` after `nbdev_pypi`, and it should handle all details, including upload.

One question  - `nbdev_conda` dumps the content of README.md into the description field of `meta.yaml`. From what I see on conda-forge, packages don't tend to have a description at all (only a summary). Should I stick to what `nbdec_conda` is doing or discourage it?

**Comment by xl0** [comment]
I've generated the yamls using `nbdev`. Don't think I can update this PR, as it was not authored by me.

@sugatoray  @BastianZim @xhochy , let's move this to PR #21332 ?

**Comment by sugatoray** [comment]
> Hii. Sorry for the delay, was busy. I'm looking into this now.
> 
> 
> 
> `lovely-tensors` now depends on `lovely-numpy`. Does it mean we should add -numpy first, then -tensors?
> 
> 
> 
> For the workflow, as I understand it, I'll have to run `nbdev_conda` after `nbdev_pypi`, and it should handle all details, including upload.
> 
> 
> 
> One question  - `nbdev_conda` dumps the content of README.md into the description field of `meta.yaml`. From what I see on conda-forge, packages don't tend to have a description at all (only a summary). Should I stick to what `nbdec_conda` is doing or discourage it?
> 
> 
> 
> 
> 
> 
> 
> 

Hi @xl0 ,

I will add lovely-numpy as a dependency to lovely-tensors.

**Comment by sugatoray** [comment]
@xhochy Could you please take a look at the PR? I believe it should be ready for merging.

Thank you.

cc: @BastianZim

**Comment by xl0** [comment]
@sugatoray, `lovely-numpy` is now in conda-forge. Would you like to update `lovely-tensors` to the latest version?

**Comment by sugatoray** [comment]
@xl0 As the current version of the recipe works properly, I would like to keep it as is. Once it gets merged, I will have an opportunity of showing you what it typically entails in managing a conda-forge recipe, when I update the version to the latest one.

**Comment by xl0** [comment]
Cool. Yeah, it makes sense.

By the way, I made this guide on conda-forge. Could you check if I missed anything important?
https://forums.fast.ai/t/getting-your-library-into-conda-forge/102860

**Comment by BastianZim** [comment]
Just one thing, we recommend running `grayskull pypi --strict-conda-forge lovely-numpy`, which generates a CF compliant (mostly) speck. Also, greyskull works but the original package is grayskull.

**Comment by xl0** [comment]
@BastianZim, thank you! I made the correction.

Do you know if anything needs to be done to merge this PR?

**Comment by BastianZim** [comment]
All good!

**Comment by BastianZim** [comment]
@conda-forge-admin, please lint

**Pull request merged**